# Exercise 2
Add the specified code for each code cell, running the cells _in order_.

Open the included **`data/words_small.txt`** file (containing a "small" list of English words) and `read()` it into a single string variable **`words_str`**. Then print the first 100 _characters_ of the `words_str` variable (use a slice operator!)

In [5]:
with open("data/words_small.txt") as my_file:
    words_str = my_file.read()
print(words_str[:100])    

abbreviate
abbreviation
abdomen
able
aboard
abroad
absent
absolute
abuse
academic
academy
accelerate


**Using a loop**, count the number of times the newline character `\n` appears in the `words_str`. This should tell you the number of words in the file (plus or minus one). You can check your work with the built-in `count()` string method (but use the loop for practice!)

In [12]:
count = 0
for char in words_str:
    if char == '\n':
        count += 1
print(count)        
print(words_str.count('\n'))

1520
1520


Create a variable **`letter_counts`** that is a dictionary whose keys are characters (letters) and whose values are the number of times that character appears in `words_str`. Print the dictionary after you make it.
- To do this, start with an empty dictionary, then loop through the letters one at a time. At each letter, increase the value associated with that letter key by one.
- Watch out for if the letter is not in the dictionary yet!
- Also _delete_ the entry for the newline `\n` character

In [15]:
letter_counts = {}
for key in words_str:
    if key == '\n':
        continue
    if letter_counts.get(key, 0) == 0: 
        letter_counts[key] = 1
    else:
        letter_counts[key] += 1
print(letter_counts)    

{'a': 703, 'b': 188, 'r': 642, 'e': 1118, 'v': 85, 'i': 515, 't': 604, 'o': 598, 'n': 550, 'd': 314, 'm': 223, 'l': 457, 's': 504, 'u': 279, 'c': 328, 'y': 189, 'p': 224, 'h': 295, 'f': 155, 'w': 161, 'g': 213, 'k': 126, 'z': 15, 'x': 23, 'j': 16, 'q': 9}


Print out all the count in the `letter_counts` dictionary in the format `"A appears 703 times"`, each on a new line, in alphabetical order. _Hint_: iterate through the sorted keys!

In [24]:
keys = sorted(letter_counts.keys())
print(keys)

for key in keys:
    print(key.upper(), "appears", letter_counts[key], "times")

['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
A appears 703 times
B appears 188 times
C appears 328 times
D appears 314 times
E appears 1118 times
F appears 155 times
G appears 213 times
H appears 295 times
I appears 515 times
J appears 16 times
K appears 126 times
L appears 457 times
M appears 223 times
N appears 550 times
O appears 598 times
P appears 224 times
Q appears 9 times
R appears 642 times
S appears 504 times
T appears 604 times
U appears 279 times
V appears 85 times
W appears 161 times
X appears 23 times
Y appears 189 times
Z appears 15 times


Print which letter from the `letter_counts` dictionary has a **value** of `515`. _Hint:_ use a loop to look through all the _items_ until you find the right one, then print that!

In [28]:
for key in keys:
    if letter_counts[key] == 515:
        print(key.upper())

I


Create a variable **`count_to_letter_dict`** that is a dictionary whose **keys** are the counts and whose **values** are the letters that have that count. This is a _reverse lookup_ dictionary. Hint: ceate a new dictionary, then loop through items in `letter_counts` and add an item to your new dictionary for each element. You can assume no duplicated counts.

In [27]:
count_to_letter_dict = {}
for key in keys:
    value = letter_counts[key]
    count_to_letter_dict[value] = key
print(count_to_letter_dict)

{703: 'a', 188: 'b', 328: 'c', 314: 'd', 1118: 'e', 155: 'f', 213: 'g', 295: 'h', 515: 'i', 16: 'j', 126: 'k', 457: 'l', 223: 'm', 550: 'n', 598: 'o', 224: 'p', 9: 'q', 642: 'r', 504: 's', 604: 't', 279: 'u', 85: 'v', 161: 'w', 23: 'x', 189: 'y', 15: 'z'}


Print the **10** most common letters and their counts with the format `"A appears 703 times"`. _Hint:_ you'll need to sort the keys to `count_to_letter_dict` and take the first 10 of those.

In [61]:
dict = sorted(count_to_letter_dict)
for i in dict[-10:]:
    print(count_to_letter_dict[i], 'appears', i, 'times')
print(dict[-10:])

c appears 328 times
l appears 457 times
s appears 504 times
i appears 515 times
n appears 550 times
o appears 598 times
t appears 604 times
r appears 642 times
a appears 703 times
e appears 1118 times
[328, 457, 504, 515, 550, 598, 604, 642, 703, 1118]


Bonus: repeat your analysis on the `data/words_huge.txt` file! Are the top letters the same?